File: src/main/java/com/manabreakstudios/discourse/ui/core/editor/NodeUI.java
package com.manabreakstudios.discourse.ui.core.editor;

import lombok.Getter;
import lombok.Setter;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;

import static com.manabreakstudios.discourse.ui.Theme.theme;

/**
 * A NodeUI is a UI component that holds a node.
 */
public class NodeUI extends JLayeredPane {

    private final NodeContent content;

    private final List<SlotComponent> slotComponentList = new ArrayList<>();

    private final Header header;

    @Getter @Setter
    private boolean isSelected = false;

    public NodeUI(NodeContent nodeContent) {
        this.content = nodeContent;
        this.header = new Header(content);

        setLayout(new ContentLayout());
        add(this.content.getContent(), JLayeredPane.DEFAULT_LAYER);
        add(header, JLayeredPane.DEFAULT_LAYER);

        for (Slot slot : this.content.getSlots()) {
            SlotComponent slotComponent = new SlotComponent(new SlotBinding(this, slot));
            add(slotComponent, JLayeredPane.POPUP_LAYER);
            slotComponentList.add(slotComponent);
        }

        setPreferredSize(this.content.getPreferredSize());
    }

    public SlotBinding getSlot(int index) {
        return new SlotBinding(this, content.getSlots().get(index));
    }

    @Override
    protected void paintComponent(Graphics g) {
        int inset = getInset();

        Graphics2D g2d = (Graphics2D) g;
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        g2d.setColor(theme().getNodeBorderColor());
        g2d.fillRoundRect(inset, inset, getWidth() - 2 * inset, getHeight() - 2 * inset, 6, 6);

        g2d.setColor(theme().getNodeSpecularColor());
        g2d.fillRoundRect(inset + 2, inset + 2, getWidth() - 2 * inset - 4, getHeight() - 2 * inset - 4, 4, 4);

        g2d.setColor(theme().getNodeBackgroundColor());
        g2d.fillRoundRect(inset + 2, inset + 3, getWidth() - 2 * inset - 4, getHeight() - 2 * inset - 5, 4, 4);

        Font font = (Font) UIManager.getDefaults().get("Font.OpenSans-ExtraBold");

        g2d.setColor(content.getColor());
        g2d.setFont(font.deriveFont(12.0f));
        g2d.drawString(content.getTitle(), inset + 10, inset + 18);
    }

    private int getInset() {
        return theme().getSlotSize() / 2;
    }

    public class Header extends JPanel {
        Header(NodeContent nodeContent) {
            setLayout(new BorderLayout());
            setOpaque(false);
            setMinimumSize(new Dimension(100, 20));
            setPreferredSize(new Dimension(1000, 20));
            setMaximumSize(new Dimension(2000, 20));
        }
    }

    public Rectangle getHitbox() {
        return getHitbox(new Rectangle());
    }

    public Rectangle getHitbox(Rectangle rv) {
        rv.setRect(getX() + getInset(), getY() + getInset(), getWidth() - 2 * getInset(), getHeight() - 2 * getInset());
        return rv;
    }

    public class ContentLayout implements LayoutManager {

        @Override
        public void addLayoutComponent(String name, Component comp) {
        }

        @Override
        public void removeLayoutComponent(Component comp) {
        }

        @Override
        public Dimension preferredLayoutSize(Container parent) {
            return content.getPreferredSize();
        }

        @Override
        public Dimension minimumLayoutSize(Container parent) {
            return preferredLayoutSize(parent);
        }

        @Override
        public void layoutContainer(Container parent) {
            int inset = getInset();
            header.setBounds(inset, inset, getWidth() - 2 * inset, 20);
            content.getContent().setBounds(inset, inset + 20, getWidth() - 2 * inset, getHeight() - 2 * inset - 20);
            for (SlotComponent slotComponent : slotComponentList) {
                int positionX = slotComponent.getSlot().getDirection() == Slot.Direction.INPUT ? 0 : (getWidth() - theme().getSlotSize());
                slotComponent.setBounds(new Rectangle(positionX, slotComponent.getSlot().getPosition(), theme().getSlotSize(), theme().getSlotSize()));
            }
        }
    }
}

File: node-editor/src/com/manabreak/node_editor/NodeUI.kt
package com.manabreak.node_editor

import com.manabreak.node_editor.Theme.Companion.theme
import java.awt.*
import java.util.*
import javax.swing.JLayeredPane
import javax.swing.JPanel

class NodeUI constructor(private val content: NodeContent) : JLayeredPane() {

    var selected = false

    private val slotComponentList = ArrayList<SlotComponent>()

    private val header = Header(content)

    override fun paintComponent(g: Graphics) {
        val inset = getInset()

        val g2d = g as Graphics2D
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)

        g2d.color = theme.nodeBorderColor
        g2d.fillRoundRect(inset, inset, width - 2 * inset, height - 2 * inset, 6, 6)

        g2d.color = theme.nodeSpecularColor
        g2d.fillRoundRect(inset + 2, inset + 2, width - 2 * inset - 4, height - 2 * inset - 4, 4, 4)

        g2d.color = theme.nodeBackgroundColor
        g2d.fillRoundRect(inset + 2, inset + 3, width - 2 * inset - 4, height - 2 * inset - 5, 4, 4)

        val font = theme.boldFont

        g2d.color = content.color
        g2d.font = font.deriveFont(12.0f)
        g2d.drawString(content.title, inset + 10, inset + 18)
    }

    init {
        layout = ContentLayout()
        add(this.content.content, JLayeredPane.DEFAULT_LAYER)
        add(header, JLayeredPane.DEFAULT_LAYER)

        for (slot in this.content.slots) {
            val slotComponent = SlotComponent(SlotBinding(this, slot))
            add(slotComponent, JLayeredPane.POPUP_LAYER)
            slotComponentList.add(slotComponent)
        }

        preferredSize = this.content.preferredSize
    }

    fun getSlot(index: Int): SlotBinding {
        return SlotBinding(this, content.slots[index])
    }

    fun getHitbox(): Rectangle {
        return getHitbox(Rectangle())
    }

    fun getHitbox(rv: Rectangle): Rectangle {
        rv.setRect((x + getInset()).toDouble(), (y + getInset()).toDouble(), (width - 2 * getInset()).toDouble(), (height - 2 * getInset()).toDouble())
        return rv
    }

    private fun getInset(): Int {
        return theme.slotSize / 2
    }

    inner class Header internal constructor(nodeContent: NodeContent) : JPanel() {
        init {
            layout = BorderLayout()
            isOpaque = false
            minimumSize = Dimension(100, 20)
            preferredSize = Dimension(1000, 20)
            maximumSize = Dimension(2000, 20)
        }
    }


    inner class ContentLayout : LayoutManager {

        override fun addLayoutComponent(name: String, comp: Component) {
        }

        override fun removeLayoutComponent(comp: Component) {
        }

        override fun preferredLayoutSize(parent: Container): Dimension {
            return content.preferredSize
        }

        override fun minimumLayoutSize(parent: Container): Dimension {
            return preferredLayoutSize(parent)
        }

        override fun layoutContainer(parent: Container) {
            val inset = getInset()
            header.setBounds(inset, inset, width - 2 * inset, 20)
            content.content.setBounds(inset, inset + 20, width - 2 * inset, height - 2 * inset - 20)
            for (slotComponent in slotComponentList) {
                val positionX = if (slotComponent.slot.direction === Slot.Direction.INPUT) 0 else width - theme.slotSize
                slotComponent.bounds = Rectangle(positionX, slotComponent.slot.position, theme.slotSize, theme.slotSize)
            }
        }
    }
}